server/network/network.js

```diff
@@ -55,67 +47,87 @@ class ClientInfo {
+/**
+ * Main entry point for networking.
+ * Initializes the networking layer, given an httpserver instance.
+ */
+export function init(server) {
   io = socketio(server);
 
-  io.on('connection', function(socket) {
-    installAlwaysAvailableHandlers(socket);
-
-    var id = new URL(socket.request.url, 'http://localhost/').searchParams.get('id');
-    if (id) {
-      // Ignore this, as it's not a main connection.
-      return;
-    }
-
-    debug('New client:', socket.id);
-    installDisplayClientHandlers(socket);
-    // Tell the clients the whole wall geo.
-    socket.emit('config', {});
+  io.on('connection', socket => {
+    // If any client asks for the current time, we tell them.
+    socket.on('time', () => {
+      socket.emit('time', now());
+    });
+    // When the client boots, it sends a start message that includes the rect
+    // of the client. We listen for that message and register that client info.
+    socket.on('client-start', config => {
+      const clientRect = Rectangle.deserialize(config.rect);
+      if (!clientRect) {
+        debug(`Bad client configuration: `, config);
+        // Close the connection with this client.
+        socket.disconnect(true);
+        return;
+      }
+      const client = new ClientInfo(clientRect, socket);
+      if (monitor.isEnabled()) {
+        monitor.update({layout: {
+          time: now(),
+          event: `newClient: ${client.rect.serialize()}`,
+        }});
+      }
+      clients[client.socket.id] = client;
+      debug(`New display: ${client.rect.serialize()}`);
+      emitter.emit('new-client', client);
+
+      socket.emit('time', now());
+    });
+
+    // When the client disconnects, we tell our listeners that we lost the client.
+    socket.once('disconnect', function() {
+      const {id} = socket;
+      if (id in clients) {
+        const {rect} = clients[id];
+        if (monitor.isEnabled()) {
+          monitor.update({layout: {
+            time: now(),
+            event: `dropClient: ${rect.serialize()}`,
+          }});
+        }
+        debug(`Lost display: ${rect.serialize()}`);
+        emitter.emit('lost-client', clients[id]);
+      } else {
+        if (monitor.isEnabled()) {
+          monitor.update({layout: {
+            time: now(),
+            event: `dropClient: id ${id}`,
+          }});
+        }
+      }
+      delete clients[id];
+    });
+
+    // If the client notices an exception, it can send us that information to
+    // the server via this channel. The framework might choose to respond to
+    // this by, say, moving on to the next module.
+    socket.on('record-error', function(e) {
+      logClientError(e);
+    });
   });
-};
 
-// Closes all communication.
-network.close = function() {
-  io.close();
-  io = undefined;
-};
-
-// Sends a message to all clients.
-network.broadcast = function(msg, data) {
-  io.emit(msg, data);
-};
+  // Set up a timer to send the current time to clients every 10 seconds.
```

Comment:
I still don't get this... I mean, just sync the clocks with ntp, yo?
I can set that up.
Or is this for some other reason?

Reply:
I just want time to be monotonically increasing, and only performance.now / process.hrtime have such guarantees. The problem is that they don't share a zero. I *could* implement this as a single bit of data in the handshake, I suppose.